assets/javascripts/kitten/karl/images/cropping-image.js

```diff
@@ -0,0 +1,275 @@
+
+    this.handleSliderChange = this.handleSliderChange.bind(this)
+    this.handleSliderAction = this.handleSliderAction.bind(this)
+
+    this.handleReady = this.handleReady.bind(this)
+    this.handleCrop = this.handleCrop.bind(this)
+
+    this.renderError = this.renderError.bind(this)
+    this.setCropperHeight = this.setCropperHeight.bind(this)
+  }
+
+  initialProps() {
+    return {
+      imageSrc: this.props.imageSrc,
+      imageCropSrc: null,
+      touched: false,
+      sliderValue: 0,
+      sliderMin: this.props.sliderMin,
+      sliderMax: this.props.sliderMax,
+    }
+  }
+
+  componentDidMount() {
+    this.setCropperHeight()
+
+    window.addEventListener('resize', this.setCropperHeight)
+  }
+
+  handleUploaderChange(file) {
+    this.setState({
+      imageSrc: file ? file : null,
+      imageCropSrc: null,
+      sliderValue: 0,
+    })
+  }
+
+  handleUploaderError(hasError) {
+    this.setState({
+      hasErrorOnUploader: hasError,
+    })
+
+    if (this.state.hasErrorOnUploader) {
+      this.setState(this.initialProps())
+    }
+  }
+
+  handleSliderChange(value) {
+    this.setState({
+      sliderValue: value,
+    })
+
+    this.refs.cropper.zoomTo(value / 100)
+  }
+
+  handleSliderAction() {
+    this.setState({ touched: true })
+  }
+
+  // Calculate the right range for the zoom slider.
+  handleReady() {
+    const imageData = this.refs.cropper.getImageData()
+    const naturalWidth = imageData.naturalWidth
+    const width = imageData.width
+    const ratio = width / naturalWidth * 100
+    const min = this.props.sliderMin + ratio
+    const max = this.props.sliderMax + ratio
+
+    this.setState({
+      sliderMin: min,
+      sliderMax: max,
+      sliderValue: min,
+    })
+  }
+
+  handleCrop() {
+    if (this.state.imageSrc) {
+      const croppedCanvas = this.refs.cropper.getCroppedCanvas()
+
+      if (croppedCanvas) {
+        this.setState({
+          imageCropSrc: croppedCanvas.toDataURL(),
+        })
+      }
+    }
+  }
+
+  setCropperHeight() {
+    const width = domElementHelper.getComputedWidth(this.refs.cropperContainer)
+    const height = width * 9 / 16
+
+    this.setState({
+      cropperHeight: height,
+    })
+  }
+
+  renderCropper() {
+    const cropperProps = {
+      ref: 'cropper',
+      className: 'k-Cropper',
+      src: this.state.imageSrc,
+      style: { height: this.state.cropperHeight },
+      aspectRatio: 16/9,
+      viewMode: 3,
+      guides: false,
+      modal: false,
+      autoCropArea: 1,
+      cropBoxMovable: false,
+      cropBoxResizable: false,
+      toggleDragModeOnDblclick: false,
+      zoomOnTouch: false,
+      zoomOnWheel: false,
+      dragMode: 'move',
+      crop: this.handleCrop,
+      ready: this.handleReady,
+    }
+
+    return (
+      <Marger key="cropper" top="2" bottom="2">
+        <div ref="cropperContainer">
+          <Cropper
+            key={ `cropper-${ this.state.cropperHeight }` }
```

Comment:
Peut-être mettre un petit commentaire pour expliquer la magie derrière ce `key`. Genre "This helps unmount and remount a new cropper to keep the component responsive".